R/utils.R

```diff
@@ -21,3 +21,129 @@
+    annotation.colors.c <- .make_heatmap_colors_continuous()
+    
+    # Initiate variables
+    next.color.index.discrete <- 1
+    next.color.index.numeric <- 1
+    
+    col_colors <- NULL
+    row_colors <- NULL
+    
+    # Columns First (if there)
+    if (!is.null(args$annotation_col)) {
+      
+        annots_numerical <- NULL
+        for (i in seq_len(ncol(args$annotation_col))){
+            
+            # Determine the distinct contents of the first annotation
+            in.this.annot <- levels(as.factor(args$annotation_col[,i]))
+            
+            # Make new colors
+            if(!is.numeric(args$annotation_col[,i])){
+                # Take colors for each, and name them.
+                new.colors <- annotation.colors.d[
+                    seq_along(in.this.annot) + next.color.index.discrete - 1
+                    ]
+                names(new.colors) <- in.this.annot
+                
+                next.color.index.discrete <-
+                    next.color.index.discrete + length(in.this.annot)
+            } else {
+                # Make a 100 color split as in pheatmap code.
+                a <- cut(args$annotation_col[,i], breaks = 100)
+                # Assign to colors.
+                this.ramp <- annotation.colors.c[next.color.index.numeric]
+                new.colors <-
+                    grDevices::colorRampPalette(c("white",this.ramp))(100)[a]
+                
+                next.color.index.numeric <- next.color.index.numeric + 1
+            }
+            # Add the new colors as the list
+            col_colors <- c(
+                col_colors,
+                list(new.colors))
+        }
+        names(col_colors) <- names(args$annotation_col)
+    }
+    
+    # Rows Second (if there)
+    if (!is.null(args$annotation_row)) {
+      
+        annots_numerical <- NULL
+        for (i in seq_len(ncol(args$annotation_row))){
+            
+            # Determine the distinct contents of the first annotation
+            in.this.annot <- levels(as.factor(args$annotation_row[,i]))
+            
+            # Make new colors
+            if(!is.numeric(args$annotation_row[,i])){
+                # Take colors for each, and name them.
+                new.colors <- annotation.colors.d[
+                    seq_along(in.this.annot) + next.color.index.discrete - 1
+                    ]
+                names(new.colors) <- in.this.annot
+                
+                next.color.index.discrete <-
+                    next.color.index.discrete + length(in.this.annot)
+            } else {
+                # Make a 100 color split as in pheatmap code.
+                a <- cut(args$annotation_row[,i], breaks = 100)
+                # Assign to colors.
+                this.ramp <- annotation.colors.c[next.color.index.numeric]
+                new.colors <-
+                    grDevices::colorRampPalette(c("white",this.ramp))(100)[a]
```

Comment:
Note: grDevices / colorRampPalette are used by the pheatmap function internally, so this isn't really an extra dependency.